base/strings/basic.jl

```diff
@@ -1,57 +1,188 @@
 
-Symbol(s::AbstractString) = Symbol(String(s))
+## required string functions ##
 
-# string types are convertible
-convert(::Type{T}, s::T) where {T<:AbstractString} = s
-convert(::Type{T}, s::AbstractString) where {T<:AbstractString} = T(s)
+"""
+    ncodeunits(s::AbstractString) -> Int
 
-## generic supplied functions ##
+Return the number of code units in a string. Indices that are in bounds to
+access this string must satisfy `1 ≤ i ≤ ncodeunits(s)`. Not all such indices
+are valid – they may not be the start of a character, but they will return a
+code unit value when calling `codeunit(s,i)`.
 
-start(s::AbstractString) = 1
-done(s::AbstractString,i) = (i > endof(s))
-getindex(s::AbstractString, i::Int) = next(s,i)[1]
-getindex(s::AbstractString, i::Integer) = s[Int(i)]
-getindex(s::AbstractString, i::Colon) = s
-getindex(s::AbstractString, r::UnitRange{<:Integer}) = s[Int(first(r)):Int(last(r))]
-# TODO: handle other ranges with stride ±1 specially?
-getindex(s::AbstractString, v::AbstractVector{<:Integer}) =
-    sprint(length(v), io->(for i in v; write(io,s[i]) end))
-getindex(s::AbstractString, v::AbstractVector{Bool}) =
-    throw(ArgumentError("logical indexing not supported for strings"))
+See also: `codeunit`, `checkbounds`, `sizeof`, `length`, `endof`
+"""
+ncodeunits(s::AbstractString)
 
-get(s::AbstractString, i::Integer, default) = isvalid(s,i) ? s[i] : default
+"""
+    codeunit(s::AbstractString) -> Type{<:Union{UInt8, UInt16, UInt32}}
+
+Return the code unit type of the given string object. For ASCII, Latin-1, or
+UTF-8 encoded strings, this would be `UInt8`; for UCS-2 and UTF-16 it would be
+`UInt16`; for UTF-32 it would be `UInt32`. The unit code type need not be
+limited to these three types, but it's hard to think of widely used string
+encodings that don't use one of these units. `codeunit(s)` is the same as
+`typeof(codeunit(s,1))` when `s` is a non-empty string.
 
+See also: `ncodeunits`
 """
-    sizeof(s::AbstractString)
+codeunit(s::AbstractString)
+
+"""
+    codeunit(s::AbstractString, i::Integer) -> Union{UInt8, UInt16, UInt32}
+
+Return the code unit value in the string `s` at index `i`. Note that
 
-The number of bytes in string `s`.
+    codeunit(s, i) :: codeunit(s)
+
+I.e. the value returned by `codeunit(s, i)` is of the type returned by
+`codeunit(s)`.
+
+See also: `ncodeunits`, `checkbounds`
+"""
+codeunit(s::AbstractString, i::Integer) = typeof(i) === Int ?
+    throw(MethodError(codeunit, Tuple{typeof(s),Int})) :
+        codeunit(s, Int(i))
+
+"""
+    isvalid(s::AbstractString, i::Integer) -> Bool
+
+Predicate indicating whether the given index is the start of the encoding of
+a character in `s` or not. If `isvalid(s, i)` is true then `s[i]` will return
+the character whose encoding starts at that index, if it's false, then `s[i]`
+will raise an invalid index error. Behavior of `next(s, i)` is similar except
+that the character is returned along with the index of the following character.
```

Comment:
I think there's a fair amount of generic string code in Base (and presumably elsewhere) that assumes that the string iteration state is the same as the string index. If it's not safe to assume that then we need to make sure that people don't use the iterator state in `s[i]`.

Reply:
I would very much like the ability to override string iteration for MyStringType (which I believe is currently possible?), and have the docs reflect the default behaviour of `next` if you don't also override `start` and `done`.

Reply:
@iamed2, I'm not quite sure what you're asking for – do you want to allow the iteration state not to be the string index? If so, I'm totally fine with that and it's mostly a documentation issue; for the built-in string types they happen to coincide. Might be worth changing the iteration state of `SubString` to match its parent's iteration state just to shake out the places where we currently rely on these being the same though.

Reply:
This'll need to be updated in that case to provide whatever is most efficient:
https://github.com/JuliaLang/julia/blob/224f1e9f8666545a5e3e0a7be721231df88e0788/base/io.jl#L649-L650
Pretty easy just to add the offset as needed, if that's what we want. At the time, I had seen that in a number of other places we define the bounds of the strings indices to be `1:n`, so just went with it:
https://github.com/JuliaLang/julia/blob/224f1e9f8666545a5e3e0a7be721231df88e0788/base/strings/basic.jl#L24-L25

Reply:
Yes, those are sane fallback definitions. Most strings work that way. The question is just that _if_ someone overrides those, do we support it working. Note that this is not actually pressing for 1.0 since if we don't allow it now, we can always allow it in the future. However, changing the iteration state of substrings might break people's code, so we would want to do that now.

Reply:
> @iamed2, I'm not quite sure what you're asking for – do you want to allow the iteration state not to be the string index? If so, I'm totally fine with that and it's mostly a documentation issue; for the built-in string types they happen to coincide. Might be worth changing the iteration state of SubString to match its parent's iteration state just to shake out the places where we currently rely on these being the same though.

Yes, exactly :)